src/db/add_package.rs

```diff
@@ -328,31 +327,103 @@ fn add_authors_into_database(
     Ok(())
 }
 
+pub(crate) fn update_crate_data_in_database(
+    conn: &Connection,
+    name: &str,
+    registry_data: &CrateData,
+) -> Result<()> {
+    info!("Updating crate data for {}", name);
+    let crate_id = conn
+        .query("SELECT id FROM crates WHERE crates.name = $1", &[&name])?
+        .get(0)
+        .get(0);
+
+    update_owners_in_database(conn, &registry_data.owners, crate_id)?;
+
+    Ok(())
+}
+
 /// Adds owners into database
-fn add_owners_into_database(conn: &Connection, owners: &[CrateOwner], crate_id: i32) -> Result<()> {
+fn update_owners_in_database(
+    conn: &Connection,
+    owners: &[CrateOwner],
+    crate_id: i32,
+) -> Result<()> {
+    let existing_owners: Vec<String> = conn
+        .query(
+            "
+        SELECT login
+        FROM owners
+        INNER JOIN owner_rels
+            ON owner_rels.oid = owners.id
+        WHERE owner_rels.cid = $1
+    ",
+            &[&crate_id],
+        )?
+        .into_iter()
+        .map(|row| row.get(0))
+        .collect();
+
     for owner in owners {
+        debug!("Updating owner data for {}: {:?}", owner.login, owner);
+
+        // Update any existing owner data since it is mutable and could have changed since last
+        // time we pulled it
         let owner_id: i32 = {
-            let rows = conn.query("SELECT id FROM owners WHERE login = $1", &[&owner.login])?;
-            if !rows.is_empty() {
-                rows.get(0).get(0)
-            } else {
-                conn.query(
-                    "INSERT INTO owners (login, avatar, name, email)
-                                 VALUES ($1, $2, $3, $4)
-                                 RETURNING id",
-                    &[&owner.login, &owner.avatar, &owner.name, &owner.email],
-                )?
-                .get(0)
-                .get(0)
-            }
+            conn.query(
+                "
+                INSERT INTO owners (login, avatar, name, email)
+                VALUES ($1, $2, $3, $4)
+                ON CONFLICT (login) DO UPDATE
+                    SET
+                        avatar = $2,
+                        name = $3,
+                        email = $4
+                RETURNING id
+            ",
+                &[&owner.login, &owner.avatar, &owner.name, &owner.email],
+            )?
+            .get(0)
+            .get(0)
         };
 
         // add relationship
-        let _ = conn.query(
+        let updated = conn.query(
             "INSERT INTO owner_rels (cid, oid) VALUES ($1, $2)",
```

Comment:
Maybe this should use `ON CONFLICT DO NOTHING`? Not sure if that would let you still debug that it was pre-existing though.

Reply:
Hmm, maybe with `RETURNING cid` it will either return one or zero rows, I'll try it.

Reply:
Didn't see a way to get the debug message, but it doesn't matter too much.